ext/ox/gen_load.c

```diff
@@ -105,24 +105,56 @@ create_prolog_doc(PInfo pi, const char *target, Attr attrs) {
 	} else if (Yes == pi->options->sym_keys) {
 #if HAVE_RB_ENC_ASSOCIATE
 	    if (0 != pi->options->rb_enc) {
+#if HAVE_RB_ENC_INTERNED_STR_CSTR
+		VALUE	rstr = rb_enc_interned_str_cstr(attrs->name, pi->options->rb_enc);
+#else
 		VALUE	rstr = rb_str_new2(attrs->name);
 
 		rb_enc_associate(rstr, pi->options->rb_enc);
+#endif
 		sym = rb_funcall(rstr, ox_to_sym_id, 0);
 	    } else {
 		sym = ID2SYM(rb_intern(attrs->name));
 	    }
 	    sym = ID2SYM(rb_intern(attrs->name));
+#endif  // HAVE_RB_ENC_ASSOCIATE
+#if HAVE_RB_INTERNED_STR_CSTR
+	    if (Yes == pi->options->intern_strings) {
+		rb_hash_aset(ah, sym, rb_interned_str_cstr(attrs->value));
+	    } else {
 #endif
 	    rb_hash_aset(ah, sym, rb_str_new2(attrs->value));
+#if HAVE_RB_INTERNED_STR_CSTR
+	    }
+#endif
 	} else {
 #if HAVE_RB_ENC_ASSOCIATE
+#if HAVE_ENC_RB_INTERNED_STR_CSTR && HAVE_RB_INTERNED_STR_CSTR
+	    if (Yes == pi->options->intern_strings) {
+	        if (0 != pi->options->rb_enc) {
+	            volatile VALUE  rstr = rb_enc_interned_str_cstr(attrs->name, pi->options->rb_enc);
+	        } else {
+	            volatile VALUE  rstr = rb_interned_str_cstr(attrs->name);
+	        }
+	    } else {
+#endif
 	    volatile VALUE	rstr = rb_str_new2(attrs->name);
 
 	    if (0 != pi->options->rb_enc) {
 		rb_enc_associate(rstr, pi->options->rb_enc);
 	    }
+#if HAVE_ENC_RB_INTERNED_STR_CSTR && HAVE_RB_INTERNED_STR_CSTR
+	    }
+#endif
+#if HAVE_RB_INTERNED_STR_CSTR
+	    if (Yes == pi->options->intern_strings) {
+	        rb_hash_aset(ah, rstr, rb_interned_str_cstr(attrs->value));
+	    } else {
+#endif
 	    rb_hash_aset(ah, rstr, rb_str_new2(attrs->value));
+#if HAVE_RB_INTERNED_STR_CSTR
+	    }
+#endif
```

Comment:
The whole section here with the various `#if` checks is next to impossible to read. I think the `volatile VALUE rstr` statements are not used in the set and would expect a compile error under the right set of conditions. Hard to tell though. How about tolerating some code duplication in favour of readability?